rest/admin-ui-ext/src/main/java/org/keycloak/admin/ui/rest/SessionsResource.java

```diff
@@ -64,49 +63,125 @@ public Stream<SessionRepresentation> realmSessions(@QueryParam("type") @DefaultV
-        return rep;
+            if (user != null) {
+                result = Stream.empty();
+                if (type == ALL || type == REGULAR) {
+                    result = Stream.concat(result, session.sessions().getUserSessionsStream(realm, user)
+                            .map(s -> toRepresentation(s, REGULAR)));
+                }
+                if (type == ALL || type == OFFLINE) {
+                    result = Stream.concat(result, session.sessions().getOfflineUserSessionsStream(realm, user)
+                            .map(s -> toRepresentation(s, OFFLINE)));
+                }
+            } else if (client != null) {
+                result = Stream.empty();
+                if (type == ALL || type == REGULAR) {
+                    result = Stream.concat(result, session.sessions().getUserSessionsStream(realm, client)
+                            .map(s -> toRepresentation(s, REGULAR)));
+                }
+                if (type == ALL || type == OFFLINE) {
+                    result = Stream.concat(result, session.sessions().getOfflineUserSessionsStream(realm, client, null, null)
+                            .map(s -> toRepresentation(s, OFFLINE)));
+                }
+            } else {
+                if (type == ALL || type == REGULAR) {
+                    final Map<String, Long> clientSessionStats = session.sessions().getActiveClientSessionStats(realm, false);
+                    clientIdSessiontypeStream = Stream.concat(clientIdSessiontypeStream, clientSessionStats
+                            .keySet().stream().map(i -> new ClientIdSessionType(i, REGULAR)));
+                }
+                if (type == ALL || type == OFFLINE) {
+                    clientIdSessiontypeStream = Stream.concat(clientIdSessiontypeStream, session.sessions().getActiveClientSessionStats(realm, true)
+                            .keySet().stream().map(i -> new ClientIdSessionType(i, OFFLINE)));
+                }
+                result = clientIdSessiontypeStream.flatMap((clientIdSessionType) -> {
+                    // TODO: Will not work if it is an external client IDs, see JpaUserSessionPersister#getUserSessionsCountsByClients
+                    ClientModel clientModel = realm.getClientById(clientIdSessionType.getClientId());
+                    switch (clientIdSessionType.getType()) {
+                        case REGULAR:
+                            return session.sessions().getUserSessionsStream(realm, clientModel)
+                                    .map(s -> toRepresentation(s, REGULAR));
+                        case OFFLINE:
+                            return session.sessions()
+                                    .getOfflineUserSessionsStream(realm, clientModel, null, null)
+                                    .map(s -> toRepresentation(s, OFFLINE));
+                    }
+                    return Stream.<SessionRepresentation>builder().build();
+                }).distinct();
```

Comment:
The `distinct()` could be applied on line 180 to optimize this op for filtered out entries.

Reply:
this has been reworked, I'm sticking with the original logic